src/library/scala/collection/immutable/ChampHashMap.scala

```diff
@@ -570,43 +860,69 @@ private final class BitmapIndexedMapNode[K, +V](val dataMap: Int, val nodeMap: I
 
 }
 
-private final class HashCollisionMapNode[K, +V](val originalHash: Int, val hash: Int, val content: Vector[(K, V)]) extends MapNode[K, V] {
+private final class HashCollisionMapNode[K, +V ](
+  val originalHash: Int,
+  val hash: Int,
+  var content: Vector[(K, V @uV)]
+  ) extends MapNode[K, V] {
 
   import Node._
 
-  require(content.size >= 2)
-  def size = content.size
+  require(content.length >= 2)
+
+  releaseFence()
+
+  private[immutable] def indexOf(key: K): Int = {
+    val iter = content.iterator
+    var i = 0
+    while (iter.hasNext) {
+      if (iter.next()._1 == key) return i
+      i += 1
+    }
+    -1
+  }
+
+  def size = content.length
 
   def get(key: K, originalHash: Int, hash: Int, shift: Int): Option[V] =
-    if (this.hash == hash) content.find(key == _._1).map(_._2) else None
+    if (this.hash == hash) {
+      val index = indexOf(key)
+      if (index >= 0) Some(content(index)._2) else None
+    } else None
+
   def getOrElse[V1 >: V](key: K, originalHash: Int, hash: Int, shift: Int, f: => V1): V1 = {
     if (this.hash == hash) {
-      content.find(key == _._1) match {
-        case Some(pair) => pair._2
-        case None => f
+      indexOf(key) match {
+        case -1 => f
+        case other => content(other)._2
       }
     } else f
   }
 
   override def containsKey(key: K, originalHash: Int, hash: Int, shift: Int): Boolean =
-    this.hash == hash && content.exists(key == _._1)
+    this.hash == hash && indexOf(key) >= 0
 
   def contains[V1 >: V](key: K, value: V1, hash: Int, shift: Int): Boolean =
-    this.hash == hash && content.exists(payload => key == payload._1 && (value.asInstanceOf[AnyRef] eq payload._2.asInstanceOf[AnyRef]))
+    this.hash == hash && {
+      val index = indexOf(key)
+      index >= 0 && (content(index)._2.asInstanceOf[AnyRef] eq value.asInstanceOf[AnyRef])
+    }
 
-  def updated[V1 >: V](key: K, value: V1, originalHash: Int, hash: Int, shift: Int): MapNode[K, V1] =
-    if (this.contains(key, value, hash, shift)) {
-      this
-    } else if (this.containsKey(key, originalHash, hash, shift)) {
-      val index = content.indexWhere(key == _._1)
-      val (beforeTuple, fromTuple) = content.splitAt(index)
-      val updatedContent = beforeTuple.appended(Tuple2(key, value)).appendedAll(fromTuple.drop(1))
-      new HashCollisionMapNode[K, V1](originalHash, hash, updatedContent)
+  def updated[V1 >: V](key: K, value: V1, originalHash: Int, hash: Int, shift: Int): MapNode[K, V1] = {
+    val index = indexOf(key)
+    if (index >= 0) {
+
+      if (content(index)._2.asInstanceOf[AnyRef] eq value.asInstanceOf[AnyRef]) {
```

Comment:
we should check if the key is eq here as well, in line with the general guidance on map key handling - looks like the old implementation didn't do this though